src/ethereum/spurious_dragon/vm/interpreter.py

```diff
@@ -251,3 +265,93 @@ def execute_code(message: Message, env: Environment) -> Evm:
         evm.has_erred = True
     finally:
         return evm
+
+
+def collect_touched_accounts(
+    evm: Evm, ancestor_had_error: bool = False
+) -> Iterable[Address]:
+    """
+    Collect all of the accounts that *may* need to be deleted based on
+    `EIP-161 <https://eips.ethereum.org/EIPS/eip-161>`_.
+    Checking whether they *do* need to be deleted happens in the caller.
+    See also: https://github.com/ethereum/EIPs/issues/716
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+    ancestor_had_error :
+        True if the ancestors of the evm object erred else False
+
+    Returns
+    -------
+    touched_accounts: `typing.Iterable`
+        returns all the accounts that were touched and may need to be deleted.
+    """
+    # collect the coinbase account if it was touched via zero-fee transfer
+    if (evm.message.caller == evm.env.origin) and evm.env.gas_price == 0:
+        yield evm.env.coinbase
+
+    # collect those explicitly marked for deletion
+    # ("beneficiary" is of SELFDESTRUCT)
+    for beneficiary in sorted(set(evm.accounts_to_delete.values())):
+        if evm.has_erred or ancestor_had_error:
+            # Special case to account for geth+parity bug
+            # https://github.com/ethereum/EIPs/issues/716
+            if beneficiary == THREE:
+                yield beneficiary
+            continue
+        else:
+            yield beneficiary
+
+    # collect account directly addressed
+    if evm.message.target != Bytes0(b""):
+        if evm.has_erred or ancestor_had_error:
+            # collect RIPEMD160 precompile even if ancestor evm had error.
+            # otherwise, skip collection from children of erred-out evm objects
+            if evm.message.target == THREE:
+                # mypy is a little dumb;
+                # Although we have evm.message.target != Bytes0(b""),
+                # my expects target to be Union[Bytes0, Address]
+                yield evm.message.target  # type: ignore
+        else:
+            # mypy is a little dumb;
+            # Although we have evm.message.target != Bytes0(b""),
+            # my expects target to be Union[Bytes0, Address]
+            yield evm.message.target  # type: ignore
+
+    # recurse into nested computations
+    # (even erred ones, since looking for RIPEMD160)
+    for child in evm.children:
+        yield from collect_touched_accounts(
+            child, ancestor_had_error=(evm.has_erred or ancestor_had_error)
+        )
+
+
+def collect_accounts_to_delete(
```

Comment:
I think this is perfectly sensible way to do `SELFDESTRUCT`, but you will need backport it to previous forks for consistency.

Reply:
I agree with you. 
-----
The previous forks didn't have child evm objects. Whenever an error would occur, I'd empty `accounts_to_delete`. eg. [evm.accounts_to_delete = set()](https://github.com/ethereum/execution-specs/blob/f037baf28cc7c546f9600fb1db91e71fba32ac88/src/ethereum/tangerine_whistle/vm/interpreter.py#L126) and the parent would collect the childs `accounts_to_delete` eg. [evm.accounts_to_delete.update(child_evm.accounts_to_delete)
](https://github.com/ethereum/execution-specs/blob/f037baf28cc7c546f9600fb1db91e71fba32ac88/src/ethereum/tangerine_whistle/vm/instructions/system.py#L241)

However, now that the evm objects have children, it felt silly to me that the parent stores the `accounts_to_delete` of the child. 
This also applies to other things like `logs`, `refund_counter `.
-----

In order to get a clean diff, I propose we add children to the previous forks too and then port `collect_accounts_to_delete`. Thoughts?

I can do this in a separate PR.